usr.sbin/apmd/apmd.8
.\"	$NetBSD: apmd.8,v 1.13 2000/02/02 01:55:55 enami Exp $
.\"
.\" Copyright (c) 1996 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by John Kohl.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"        This product includes software developed by the NetBSD
.\"        Foundation, Inc. and its contributors.
.\" 4. Neither the name of The NetBSD Foundation nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd September 12, 1996
.Dt APMD 8
.Os
.Sh NAME
.Nm apmd
.Nd Advanced Power Management monitor daemon
.Sh SYNOPSIS
.Nm ""
.Op Fl adlqs
.Bk -words
.Op Fl t Ar seconds
.Ek
.Bk -words
.Op Fl S Ar sockname
.Ek
.Bk -words
.Op Fl m Ar sockmode
.Ek
.Bk -words
.Op Fl o Ar sockowner:sockgroup
.Ek
.Bk -words
.Op Fl f Ar devname
.Ek
.Sh DESCRIPTION
.Nm
monitors the advanced power management (APM) pseudo-device, acting on
signaled events and upon user requests as sent by the
.Xr apm 8
utility.
For suspend and standby request events delivered by the BIOS, or via
.Xr apm 8 ,
.Nm
runs the appropriate configuration program (if one exists),
syncs the buffer cache to disk and initiates the requested mode.
When resuming after suspend or standby,
.Nm
runs the appropriate configuration utility (if one exists).
For power status change events,
.Nm
fetches the current status and reports it via
.Xr syslog 3
with logging facility
.Dv LOG_DAEMON .
.Pp
.Nm
announces the transition to standby mode with a single high tone on the
speaker (using the
.Pa /dev/speaker
device).
Suspends are announced with two high tones.
.Pp
.Nm
periodically polls the APM driver for the current power state.
If the battery charge level changes substantially or the external power
status changes, the new status is logged.  The polling rate defaults to
once per 10 minutes, but may be specified using the
.Fl t
command-line flag.
.Pp
If the
.Fl s
flag is specified, the current battery statistics are reported via
.Xr syslog 3
and
.Nm
exits without monitoring the APM status.
.Pp
If the
.Fl a
flag is specified, any BIOS-initiated suspend or standby requests are
ignored if the system is connected to line current and not running from
batteries (user requests are still honored).
.Pp
If the
.Fl l
flag is specified, then a low-battery event causes a suspend request
to occur.
.Pp
If the
.Fl d
flag is specified,
.Nm
enters debug mode, logging to facility
.Dv LOG_LOCAL1
and staying in the foreground on the controlling terminal.
.Pp
If the
.Fl q
flag is specified,
.Nm
does not announce suspend and standby requests on the speaker.
.Pp
When a client requests a suspend or stand-by mode,
.Nm
does not wait for positive confirmation that the requested
mode has been entered before replying to the client; to do so would mean
the client does not get a reply until the system resumes from its sleep state.
Rather,
.Nm
replies with the intended state to the client and then places the system
in the requested mode after running the configuration script and
flushing the buffer cache.
.Pp
Actions can be configured for the five transitions:
.Cm suspend ,
.Cm standby ,
.Cm resume ,
.Cm line
or
.Cm battery .
The suspend and standby actions are run prior to
.Nm
performing any other actions (such as disk syncs) and entering the new
mode.  The resume program is run after resuming from a stand-by or
suspended state.
.Pp
The line and battery actions are run after switching power sources to
AC (line) or battery, respectively.  The appropriate line or battery action
is also run upon the startup of apmd based on the current power source.
.Sh FILES
.Pa /etc/apm/suspend ,
.Pa /etc/apm/standby ,
.Pa /etc/apm/resume ,
.Pa /etc/apm/line
and
.Pa /etc/apm/battery
are the files that contain the host's customized actions.
Each file must be an executable binary or shell script suitable
for execution by the
.Xr execve 2
function.
If you wish to have the same program or script control all transitions, it
may determine which transition is in progress by examining its
.Va argv[0]
which is set to one of
.Ar suspend ,
.Ar standby ,
.Ar resume ,
.Ar line
or
.Ar battery .
See
.Pa /usr/share/examples/apm/script
for such an example script.
.Pp
.Pa /var/run/apmdev
is the default UNIX-domain socket used for communication with
.Xr apm 8 .
The
.Fl S
flag may be used to specify an alternate socket name.
The socket is protected by default to mode 0660, UID 0, GID 0;
the
.Fl m
and
.Fl o
flags may be used to specify different modes or owners/groups, respectively.
.Pp
.Pa /dev/apmctl
is the default device used to control the APM kernel driver.
The
.Fl f
flag may be used to specify an alternate device file name.
.Sh SEE ALSO
.Xr execve 2 ,
.Xr syslog 3 ,
.Xr apm 4 ,
.Xr speaker 4 ,
.Xr apm 8 ,
.Xr syslogd 8
.Sh REFERENCES
Advanced Power Management (APM) BIOS Interface Specification (revision
1.1), Intel Corporation and Microsoft Corporation.
.Sh HISTORY
The
.Nm
command appeared in
.Nx 1.3 .


usr.sbin/apmd/apmd.c
/*	$NetBSD: apmd.c,v 1.13 2000/01/13 16:04:53 jlam Exp $	*/

/*-
 * Copyright (c) 1996 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by John Kohl.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *        This product includes software developed by the NetBSD
 *        Foundation, Inc. and its contributors.
 * 4. Neither the name of The NetBSD Foundation nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <errno.h>
#include <syslog.h>
#include <fcntl.h>
#include <unistd.h>
#include <util.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <pwd.h>
#include <grp.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/wait.h>
#include <machine/apmvar.h>
#include <err.h>
#include "pathnames.h"
#include "apm-proto.h"

#define MAX(a,b) (a > b ? a : b)
#define TRUE 1
#define FALSE 0

const char apmdev[] = _PATH_APM_CTLDEV;
const char sockfile[] = _PATH_APM_SOCKET;

static int debug = 0;

extern char *__progname;

void usage (void);
int power_status (int fd, int force, struct apm_power_info *pinfo);
int bind_socket (const char *sn, mode_t mode, uid_t uid, gid_t gid);
enum apm_state handle_client(int sock_fd, int ctl_fd);
void suspend(int ctl_fd);
void stand_by(int ctl_fd);
void resume(int ctl_fd);
void sigexit(int signo);
void make_noise(int howmany);
void do_etc_file(const char *file);
void do_ac_state(int state);

void
sigexit(int signo)
{
    exit(1);
}

void
usage(void)
{
    fprintf(stderr,"usage: %s [-d] [-s] [-a] [-l] [-q] [-t seconds] [-S sockname]\n\t[-m sockmode] [-o sockowner:sockgroup] [-f devname]\n", __progname);
    exit(1);
}


int
power_status(int fd, int force, struct apm_power_info *pinfo)
{
    struct apm_power_info bstate;
    static struct apm_power_info last;
    int acon = 0;

    if (ioctl(fd, APM_IOC_GETPOWER, &bstate) == 0) {
	/* various conditions under which we report status:  something changed
	   enough since last report, or asked to force a print */
	if (bstate.ac_state == APM_AC_ON)
	    acon = 1;
	if (force ||
	    bstate.ac_state != last.ac_state ||
	    bstate.battery_state != last.battery_state ||
	    (bstate.minutes_left && bstate.minutes_left < 15) ||
	    abs(bstate.battery_life - last.battery_life) > 20) {
	    if (bstate.minutes_left)
		syslog(LOG_NOTICE,
		       "battery status: %s. external power status: %s. "
		       "estimated battery life %d%% (%d minutes)",
		       battstate(bstate.battery_state),
		       ac_state(bstate.ac_state), bstate.battery_life,
		       bstate.minutes_left);
	    else
		syslog(LOG_NOTICE,
		       "battery status: %s. external power status: %s. "
		       "estimated battery life %d%%",
		       battstate(bstate.battery_state),
		       ac_state(bstate.ac_state), bstate.battery_life);
	    last = bstate;
	}
	if (pinfo)
	    *pinfo = bstate;
    } else
	syslog(LOG_ERR, "cannot fetch power status: %m");
    return acon;
}

static char *socketname;

static void sockunlink(void);

static void
sockunlink(void)
{
    if (socketname)
	(void) remove(socketname);
}

int
bind_socket(const char *sockname, mode_t mode, uid_t uid, gid_t gid)
{
    int sock;
    struct sockaddr_un s_un;

    sock = socket(AF_LOCAL, SOCK_STREAM, 0);
    if (sock == -1)
	err(1, "cannot create local socket");

    s_un.sun_family = AF_LOCAL;
    strncpy(s_un.sun_path, sockname, sizeof(s_un.sun_path));
    s_un.sun_len = SUN_LEN(&s_un);
    /* remove it if present, we're moving in */
    (void) remove(sockname);
    if (bind(sock, (struct sockaddr *)&s_un, s_un.sun_len) == -1)
	err(1, "cannot create APM socket");
    if (chmod(sockname, mode) == -1 || chown(sockname, uid, gid) == -1)
	err(1, "cannot set socket mode/owner/group to %o/%d/%d",
	    mode, uid, gid);
    listen(sock, 1);
    socketname = strdup(sockname);
    atexit(sockunlink);
    return sock;
}

enum apm_state
handle_client(int sock_fd, int ctl_fd)
{
    /* accept a handle from the client, process it, then clean up */
    int cli_fd;
    struct sockaddr_un from;
    int fromlen = sizeof(from);
    struct apm_command cmd;
    struct apm_reply reply;

    cli_fd = accept(sock_fd, (struct sockaddr *)&from, &fromlen);
    if (cli_fd == -1) {
	syslog(LOG_INFO, "client accept failure: %m");
	return NORMAL;
    }
    if (recv(cli_fd, &cmd, sizeof(cmd), 0) != sizeof(cmd)) {
	(void) close(cli_fd);
	syslog(LOG_INFO, "client size botch");
	return NORMAL;
    }
    if (cmd.vno != APMD_VNO) {
	close(cli_fd);			/* terminate client */
	/* no error message, just drop it. */
	return NORMAL;
    }
    power_status(ctl_fd, 0, &reply.batterystate);
    switch (cmd.action) {
    default:
	reply.newstate = NORMAL;
	break;
    case SUSPEND:
	reply.newstate = SUSPENDING;
	break;
    case STANDBY:
	reply.newstate = STANDING_BY;
	break;
    }
    reply.vno = APMD_VNO;
    if (send(cli_fd, &reply, sizeof(reply), 0) != sizeof(reply)) {
	syslog(LOG_INFO, "client reply botch");
    }
    close(cli_fd);
    return reply.newstate;
}

static int speaker_ok = TRUE;

void
make_noise(howmany)
int howmany;
{
    int spkrfd;
    int trycnt;

    if (!speaker_ok)		/* don't bother after sticky errors */
	return;

    for (trycnt = 0; trycnt < 3; trycnt++) {
	spkrfd = open(_PATH_DEV_SPEAKER, O_WRONLY);
	if (spkrfd == -1) {
	    switch (errno) {
	    case EBUSY:
		usleep(500000);
		errno = EBUSY;
		continue;
	    case ENOENT:
	    case ENODEV:
	    case ENXIO:
	    case EPERM:
	    case EACCES:
		syslog(LOG_INFO,
		       "speaker device " _PATH_DEV_SPEAKER " unavailable: %m");
		speaker_ok = FALSE;
		return;
	    }
	} else
	    break;
    }
    if (spkrfd == -1) {
	syslog(LOG_WARNING, "cannot open " _PATH_DEV_SPEAKER ": %m");
	return;
    }
    syslog(LOG_DEBUG, "sending %d tones to speaker\n", howmany);
    write (spkrfd, "o4cc", 2 + howmany);
    close(spkrfd);
    return;
}


void
suspend(int ctl_fd)
{
    do_etc_file(_PATH_APM_ETC_SUSPEND);
    sync();
    make_noise(2);
    sync();
    sync();
    sleep(1);
    ioctl(ctl_fd, APM_IOC_SUSPEND, 0);
}

void
stand_by(int ctl_fd)
{
    do_etc_file(_PATH_APM_ETC_STANDBY);
    sync();
    make_noise(1);
    sync();
    sync();
    sleep(1);
    ioctl(ctl_fd, APM_IOC_STANDBY, 0);
}

#define TIMO (10*60)			/* 10 minutes */

void
resume(int ctl_fd)
{
    do_etc_file(_PATH_APM_ETC_RESUME);
}

int
main(int argc, char *argv[])
{
    const char *fname = apmdev;
    int ctl_fd, sock_fd, ch, ready;
    int statonly = 0;
    fd_set devfds;
    fd_set selcopy;
    struct apm_event_info apmevent;
    int suspends, standbys, resumes;
    int noacsleep = 0;
    int lowbattsleep = 0;
    mode_t mode = 0660;
    struct timeval tv = {TIMO, 0}, stv;
    const char *sockname = sockfile;
    char *user, *group;
    char *scratch;
    uid_t uid = 0;
    gid_t gid = 0;
    struct passwd *pw;
    struct group *gr;

    while ((ch = getopt(argc, argv, "qaldsf:t:S:m:o:")) != -1)
	switch(ch) {
	case 'q':
	    speaker_ok = FALSE;
	    break;
	case 'a':
	    noacsleep = 1;
	    break;
	case 'l':
	    lowbattsleep = 1;
	    break;
	case 'd':
	    debug = 1;
	    break;
	case 'f':
	    fname = optarg;
	    break;
	case 'S':
	    sockname = optarg;
	    break;
	case 't':
	    tv.tv_sec = strtoul(optarg, 0, 0);
	    if (tv.tv_sec == 0)
		usage();
	    break;
	case 'm':
	    mode = strtoul(optarg, 0, 8);
	    if (mode == 0)
		usage();
	    break;
	case 'o':
	    /* (user):(group) */
	    user = optarg;
	    group = strchr(user, ':');
	    if (group)
		*group++ = '\0';
	    if (*user) {
		uid = strtoul(user, &scratch, 0);
		if (*scratch != '\0') {
		    pw = getpwnam(user);
		    if (pw)
			uid = pw->pw_uid;
		    else
			errx(1, "user name `%s' unknown", user);
		}
	    }
	    if (group && *group) {
		gid = strtoul(group, &scratch, 0);
		if (*scratch != '\0') {
		    gr = getgrnam(group);
		    if (gr)
			gid = gr->gr_gid;
		    else
			errx(1, "group name `%s' unknown", group);
		}
	    }
	    break;
	case 's':			/* status only */
	    statonly = 1;
	    break;
	case '?':

	default:
	    usage();
	}
    argc -= optind;
    argv += optind;
    if ((ctl_fd = open(fname, O_RDWR)) == -1) {
	(void)err(1, "cannot open device file `%s'", fname);
    }
    if (debug) {
	openlog(__progname, LOG_CONS, LOG_LOCAL1);
    } else {
	openlog(__progname, LOG_CONS, LOG_DAEMON);
	setlogmask(LOG_UPTO(LOG_NOTICE));
	daemon(0, 0);
	pidfile(NULL);
    }
    if (statonly) {
        power_status(ctl_fd, 1, 0);
	exit(0);
    } else {
	struct apm_power_info pinfo;
	power_status(ctl_fd, 1, &pinfo);
	do_ac_state(pinfo.ac_state);
    }

    (void) signal(SIGTERM, sigexit);
    (void) signal(SIGHUP, sigexit);
    (void) signal(SIGINT, sigexit);
    (void) signal(SIGPIPE, SIG_IGN);


    sock_fd = bind_socket(sockname, mode, uid, gid);

    FD_ZERO(&devfds);
    FD_SET(ctl_fd, &devfds);
    FD_SET(sock_fd, &devfds);



    for (selcopy = devfds, errno = 0, stv = tv;
	 (ready = select(MAX(ctl_fd,sock_fd)+1, &selcopy, 0, 0, &stv)) >= 0 ||
	     errno == EINTR;
	 selcopy = devfds, errno = 0, stv = tv) {
	if (errno == EINTR)
	    continue;
	if (ready == 0) {
	    /* wakeup for timeout: take status */
	    power_status(ctl_fd, 0, 0);
	}
	if (FD_ISSET(ctl_fd, &selcopy)) {
	    suspends = standbys = resumes = 0;
	    while (ioctl(ctl_fd, APM_IOC_NEXTEVENT, &apmevent) == 0) {
		syslog(LOG_DEBUG, "apmevent %04x index %d", apmevent.type,
		       apmevent.index);
		switch (apmevent.type) {
		case APM_SUSPEND_REQ:
		case APM_USER_SUSPEND_REQ:
		case APM_CRIT_SUSPEND_REQ:
		    suspends++;
		    break;
		case APM_BATTERY_LOW:
		    if (lowbattsleep)
			suspends++;
		    break;
		case APM_USER_STANDBY_REQ:
		case APM_STANDBY_REQ:
		    standbys++;
		    break;
#if 0
		case APM_CANCEL:
		    suspends = standbys = 0;
		    break;
#endif
		case APM_NORMAL_RESUME:
		case APM_CRIT_RESUME:
		case APM_SYS_STANDBY_RESUME:
		    resumes++;
		    break;
		case APM_POWER_CHANGE:
		{
		    struct apm_power_info pinfo;
		    power_status(ctl_fd, 0, &pinfo);
		    do_ac_state(pinfo.ac_state);
		    break;
		}
		default:
		    break;
		}
	    }
	    if ((standbys || suspends) && noacsleep &&
		power_status(ctl_fd, 0, 0)) {
		syslog(LOG_DEBUG, "not sleeping cuz AC is connected");
	    } else if (suspends) {
		suspend(ctl_fd);
	    } else if (standbys) {
		stand_by(ctl_fd);
	    } else if (resumes) {
		resume(ctl_fd);
		syslog(LOG_NOTICE, "system resumed from APM sleep");
	    }
	    ready--;
	}
	if (ready == 0)
	    continue;
	if (FD_ISSET(sock_fd, &selcopy)) {
	    switch (handle_client(sock_fd, ctl_fd)) {
	    case NORMAL:
		break;
	    case SUSPENDING:
		suspend(ctl_fd);
		break;
	    case STANDING_BY:
		stand_by(ctl_fd);
		break;
	    }
	}
    }
    syslog(LOG_ERR, "select failed: %m");
    exit(1);
}

void
do_etc_file(const char *file)
{
    pid_t pid;
    int status;
    const char *prog;

    /* If file doesn't exist, do nothing. */
    if (access(file, X_OK|R_OK)) {
	syslog(LOG_DEBUG, "do_etc_file(): cannot access file %s", file);
	return;
    }

    prog = strrchr(file, '/');
    if (prog)
	prog++;
    else
	prog = file;

    pid = fork();
    switch (pid) {
    case -1:
	syslog(LOG_ERR, "failed to fork(): %m");
	return;
    case 0:
	/* We are the child. */
	execl(file, prog, NULL);
	_exit(-1);
	/* NOTREACHED */
    default:
	/* We are the parent. */
	wait4(pid, &status, 0, 0);
	if (WIFEXITED(status))
	    syslog(LOG_DEBUG, "%s exited with status %d", file,
		   WEXITSTATUS(status));
	else {
	    syslog(LOG_ERR, "%s exited abnormally.", file);
	}
	break;
    }
}

void
do_ac_state(int state)
{
	switch (state) {
	case APM_AC_OFF:
		do_etc_file(_PATH_APM_ETC_BATTERY);
		break;
	case APM_AC_ON:
	case APM_AC_BACKUP:
		do_etc_file(_PATH_APM_ETC_LINE);
		break;
	default:
		/* Silently ignore */
	}
}
